### PR TITLE
Bootstrap 5 migration 

### PR DESCRIPTION
**Pull request by sagargg**
Fixes #
 #6012 
### Proposed fixes:
Bootstrap  upgrade  5 

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

**Comment by amercader** [comment]
@sagargg It's so great to see progress here. How are you getting on with this? Do you need any support from the tech team or other devs?

**Comment by sagargg** [comment]
Hi @amercader, I think most of the things are done and ready for review. Currenlty  I'm checking the leftover broken item and fixing it.  Also, there are 97 tests failing now and it seems that most of them are not related to these changes.

**Comment by Zharktas** [comment]
@sagargg now the failing tests should be related to this PR, they mostly contain tests that look for specific HTML.

**Comment by pdelboca** [comment]
This is a massive work @sagargg , thanks for pulling it together!

I have been reviewing and found some things in the UI that may need a few tweaks.

1. Buttons to add new resources (or images) seems off: 
image

2. Flask-DebugToolbar div seems off too:
image

3. Home page seems to have an issue sticking the footer to the bottom of the page:
image


This errors seems minor but it worth be good to take a look at them and make an action plan to merge this before other PRs can introduce massive changes to templates.

cc @ckan/core

**Comment by pdelboca** [comment]
@sagargg it would be nice to not let die this effort and do a little bit more so we can merge it for the next patch release `2.10.0` in January.

In order to do it we need to:
- Add a mechanism to support BS3 and BS5
- Test that it's compatible to run CKAN in both
- Merge this changes into master

## How to support BS3 and BS5

We did it before when migrating BS2 to BS3. Basically we need to:
- Maintain folders for BS3 (This is already done in the PR with `templates-bs3` and `public-bs3`)
- Update [here](https://github.com/ckan/ckan/blob/9d262b5a156ff1f947618df2b261eca79f77f78d/ckan/config/environment.py#L38-L46) and [here](https://github.com/ckan/ckan/blob/9d262b5a156ff1f947618df2b261eca79f77f78d/ckan/config/environment.py#L188-L197) to allow use of BS3 templates.

## How to test for BS3 compatibility
- Add needed configurations in `ckan.ini` file: 
  - `ckan.base_public_folder=public-bs3`
  - `ckan.base_templates_folder=templates-bs3`
- Run CKAN

@sagargg from an initial test it's not working properly when trying to update to BS3. Do you have time to work on this PR between now and January? If so it would be awesome so we can add this work to `2.10.0`.

Let me know if you have time to work on this PR! (I can support you with whatever you need)

cc @ckan/core

**Comment by pdelboca** [comment]
Also it's worth considering that `ckanext-scheming` is a heavy use of `panels` that are deprecated in BS5. However since the release schedule is different from CKAN it's okay to merge BS5 if we can keep compatibility to BS3.

**Comment by sagargg** [comment]
Hi @pdelboca, thanks for providing the review feedbacks.  I made changes to support both BS3 and BS4 templates. Also fixed styling from the above comment.

**Comment by pdelboca** [comment]
Hi @sagargg ! Sorry for the delay on answering (holidays). Thanks for the update, now I can run in `bs3` and `bs5` without failing.

There are still some things to address:

### BS 5
 - The `Upload` button on the resource create page doesn't trigger any action
 
upload button


### BS3
Even when the system can run there are some components of the UI broken:
 - The navbar is not displayed correctly (Logo on the rigth and the navigation links are not render properly)
 - The organization grid is broken when uploading an Organization image
 
 The following screenshot provides an example of both the points mentioned:
 
image

**Comment by sagargg** [comment]
@pdelboca  That should be fixed now. Let me know if you found other issues.

**Comment by amercader** [comment]
@sagargg @pdelboca I've merged https://github.com/ckan/ckan/pull/6227 which will probably mean some work here to adapt it, but unless you think otherwise I think it's better to handle it in a separate PR.

(@wardi I wonder if it makes sense to add the new resource upload snippets just to the new default Boostrap 5 templates and leave the BS 3 with the current snippets. Or will that complicate things?)

So let's merge this and create a followup issue with pending tasks like the one above. To get this merged let's look at why the checks are failing and fix the conflicts.

Thanks!

**Comment by wardi** [comment]
@amercader I think it's better to have it in both because it's kind of strange to require bs5 for custom url types in resources. Is there a technical reason why this would be difficult?

**Comment by sagargg** [comment]
Hi @pdelboca  @amercader,  What is blocking here to merge this PR?

**Comment by Zharktas** [comment]
@sagargg it's mostly time constraints, I'll intend to have some final review of this hopefully next week.

**Comment by Zharktas** [comment]
When bs3-templates are used, it loads the wrong base directory, it still uses public instead of public-bs3 which results broken UI's.
Additionally the trash in ckan-admin seems to be broken on bs5 templates.

image

In above image, there should be deleted dataset shown.

image

**Comment by Zharktas** [comment]
We discussed this on the dev meeting and once the bs3 templates use bs3 javascript files, this can be merged, rest of the issues can be fixed later on as the next ckan release is not scheduled yet.

**Comment by sagargg** [comment]
@Zharktas  could you double check please. bs3 resources are loading for me. 
```
ckan.base_public_folder=public-bs3
ckan.base_templates_folder=templates-bs3
```

**Comment by sagargg** [comment]
@Zharktas  fixed broken UI in admin trash page.

**Comment by Zharktas** [comment]
Bah, there where two different settings, which I don't like very much but that's how it was previously in similar migration so no need to change that. Can you update the error messages to contain valid options [here](https://github.com/ckan/ckan/blob/fd2a5e1a0eaeb3c93215bbf87e4938ed2313af2e/ckan/config/environment.py#L42-L45) and [here](https://github.com/ckan/ckan/blob/fd2a5e1a0eaeb3c93215bbf87e4938ed2313af2e/ckan/config/environment.py#L193-L197) and add a changelog entry in `changes/6307.feature` file, then were good to go.

**Comment by sagargg** [comment]
@Zharktas  Done.

File: ckan/config/environment.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def load_environment(conf):
     if static_files not in valid_base_public_folder_names:
         raise CkanConfigurationException(
             'You provided an invalid value for ckan.base_public_folder. '
-            'Possible values are: "public".'
+            'Possible values are: "public-bs3".'
```
Comment (Zharktas) [inline ckan/config/environment.py:45]:
```suggestion
            'Possible values are: "public" and "public-bs3".'
```

File: ckan/config/environment.py [hunk 2]
```diff
@@ -193,7 +193,7 @@ def update_config():
     if templates not in valid_base_templates_folder_names:
         raise CkanConfigurationException(
             'You provided an invalid value for ckan.base_templates_folder. '
-            'Possible values are: "templates".'
+            'Possible values are: "templates-bs3".'
```
Comment (Zharktas) [inline ckan/config/environment.py:196]:
```suggestion
            'Possible values are: "templates" and "templates-bs3".'
```

**Review by Zharktas — CHANGES_REQUESTED** [review]
Just a minor thing to fix and then I'll merge this :)

**Comment by pdelboca** [comment]
This is a great work @sagargg ! Thank you so much for this PR :)

And thanks @Zharktas for following up and merging it.

**Comment by amercader** [comment]
:tada: Thanks @sagargg for the hard work and @Zharktas for the review!

**Pull request merged**